File: android/app/src/main/java/com/forge/mawaqitmasjid/models/ApiError.kt
// PR-29 — typed API error.
package com.forge.mawaqitmasjid.models

sealed class ApiError(val reason: String) : Throwable(reason) {
    data class Transport(val detail: String) : ApiError("Network error: $detail")
    object Unauthorized : ApiError("Please sign in again.")
    object NotFound : ApiError("Requested resource not found.")
    data class Server(val code: Int, val detail: String) : ApiError("Server error $code: $detail")
    data class Decoding(val detail: String) : ApiError("Response could not be decoded: $detail")
    object Offline : ApiError("No internet connection.")
}

sealed class ApiResult<out T> {
    data class Success<T>(val value: T) : ApiResult<T>()
    data class Failure(val error: ApiError) : ApiResult<Nothing>()
}

File: ios/MawaqitMasjid/Sources/Models/ApiError.swift
// PR-29 — typed API error.
import Foundation

public enum ApiError: Error, LocalizedError, Equatable {
    case transport(String)
    case unauthorized
    case notFound
    case server(Int, String)
    case decoding(String)
    case offline

    public var errorDescription: String? {
        switch self {
        case .transport(let msg): return "Network error: \(msg)"
        case .unauthorized: return "Please sign in again."
        case .notFound: return "Requested resource not found."
        case .server(let code, let msg): return "Server error \(code): \(msg)"
        case .decoding(let msg): return "Response could not be decoded: \(msg)"
        case .offline: return "No internet connection."
        }
    }
}
